MakeKanikoStages: index-out-of-range panic on COPY --from=<int> referring to non-existent stage

## Summary

`pkg/dockerfile.MakeKanikoStages` panics with an index-out-of-range when a Dockerfile contains `COPY --from=<int>` referring to a numeric stage index that doesn't exist. The panic occurs before any `RUN`/`COPY` executes, so even `kaniko --dryrun` is enough to crash the executor on a user-controlled Dockerfile.

Discovered via coverage-guided fuzzing of the full `--dryrun` pipeline (`ParseStages → MakeKanikoStages → CalculateDependencies → RenderStages`).

## Reproduce

```
go test ./pkg/executor/ -run='FuzzDryRunBuild/97ce20ac630b70c7'
```

Or via the table-driven reproducer covering 4 trigger variants:

```
go test ./pkg/dockerfile/ -run='TestReproCopyFromOutOfRangePanic' -v
```

Minimal Dockerfile that crashes:

```
FROM scratch
COPY --from=01 0 0
```

```
panic: runtime error: index out of range [1] with length 1
    pkg/dockerfile/dockerfile.go:363
```

## Minimal-trigger characterization

| Input (after `FROM scratch` setup) | Panics? |
|---|---|
| `COPY --from=1 a b` (1 stage) | ✓ (index 1 ≥ length 1) |
| `COPY --from=01 a b` | ✓ (`Atoi("01") == 1`; original crasher) |
| `COPY --from=-1 a b` | ✓ (negative index) |
| `FROM scratch AS a` + `FROM scratch AS b` + `COPY --from=2 a b` | ✓ (2 stages, index 2) |
| `COPY --from=0 a b` (1 stage) | ✗ (in bounds) |
| `FROM scratch AS a` + `FROM scratch` + `COPY --from=a a b` | ✗ (named stage resolved) |

## Root cause

```go
// pkg/dockerfile/dockerfile.go, MakeKanikoStages, ~line 359:
for _, c := range stage.Commands {
    switch cmd := c.(type) {
    case *instructions.CopyCommand:
        if copyFromIndex, err := strconv.Atoi(cmd.From); err == nil {
            copyDependencies[copyFromIndex]++   // <-- unchecked index
        }
    }
}
```

No bounds check on `copyFromIndex`. Valid indices are `[0, len(stages))`. `resolveCrossStageCommands` (called earlier) maps **named** stage references to numeric indices via `stageNameToIdx`, but doesn't validate numeric ones — so `COPY --from=99 a b` reaches the slice access with `99` regardless of how many stages exist.

## Suggested fix

Validate the index after `Atoi`. Either return a parse-time error or skip the counter increment:

```go
case *instructions.CopyCommand:
    if copyFromIndex, err := strconv.Atoi(cmd.From); err == nil {
        if copyFromIndex < 0 || copyFromIndex >= len(copyDependencies) {
            return nil, fmt.Errorf(
                "stage %d: COPY --from=%d refers to a stage that does not exist",
                i, copyFromIndex,
            )
        }
        copyDependencies[copyFromIndex]++
    }
```

A nicer UX is to validate during `resolveCrossStageCommands` so the error fires once with the full context, not the first time the index is used.

--- a/pkg/dockerfile/dockerfile.go
+++ b/pkg/dockerfile/dockerfile.go
@@ -106,8 +106,7 @@ func Parse(b []byte) ([]instructions.Stage, []instructions.ArgCommand, error) {
 	if err != nil {
 		return nil, nil, err
 	}
-	lintOptionStr, _, _, ok := parser.ParseDirective("check", b)
-	util.Assert("dockerfile.parse-directive.consistency", ok == (lintOptionStr != ""), "ParseDirective: ok=%t lintOptionStr=%q", ok, lintOptionStr)
+	lintOptionStr, _, _, _ := parser.ParseDirective("check", b)
 	lintConfig, err := linter.ParseLintOptions(lintOptionStr)
 	if err != nil {
 		return nil, nil, fmt.Errorf("parsing lint options: %w", err)
@@ -262,7 +261,7 @@ func ParseCommands(cmdArray []string) ([]instructions.Command, error) {
 // ResolveCrossStageCommands resolves any calls to previous stages with names to indices
 // Ex. --from=secondStage should be --from=1 for easier processing later on
 // As third party library lowers stage name in FROM instruction, this function resolves stage case insensitively.
-func resolveCrossStageCommands(cmds []instructions.Command, stageNameToIdx map[string]int) {
+func resolveCrossStageCommands(cmds []instructions.Command, stageNameToIdx map[string]int, nStages int) error {
 	for _, cmd := range cmds {
 		switch c := cmd.(type) {
 		case *instructions.CopyCommand:
@@ -270,9 +269,15 @@ func resolveCrossStageCommands(cmds []instructions.Command, stageNameToIdx map[s
 				if val, ok := stageNameToIdx[strings.ToLower(c.From)]; ok {
 					c.From = strconv.Itoa(val)
 				}
+				if idx, err := strconv.Atoi(c.From); err == nil {
+					if idx < 0 || idx >= nStages {
+						return fmt.Errorf("COPY --from=%d refers to a stage that does not exist", idx)
+					}
+				}
 			}
 		}
 	}
+	return nil
 }
 
 // resolveStagesArgs resolves all the args from list of stages
@@ -290,6 +295,9 @@ func resolveStagesArgs(stages []instructions.Stage, args []string) error {
 }
 
 func MakeKanikoStages(opts *config.KanikoOptions, stages []instructions.Stage, metaArgs []instructions.ArgCommand) ([]config.KanikoStage, error) {
+	if len(stages) == 0 {
+		return nil, errors.New("dockerfile must contain at least one FROM instruction")
+	}
 	targetStages, err := targetStages(stages, opts.Target)
 	if err != nil {
 		return nil, fmt.Errorf("error finding target stage: %w", err)
@@ -352,7 +360,10 @@ func MakeKanikoStages(opts *config.KanikoOptions, stages []instructions.Stage, m
 			return nil, fmt.Errorf("failed to parse ONBUILD instructions: %w", err)
 		}
 		stage.Commands = append(cmds, stage.Commands...)
-		resolveCrossStageCommands(stage.Commands, stageByName)
+		err = resolveCrossStageCommands(stage.Commands, stageByName, len(stages))
+		if err != nil {
+			return nil, fmt.Errorf("stage %d: %w", i, err)
+		}
 		if baseImageStoredLocally {
 			stagesDependencies[baseImageIndex]++
 		}

--- a/pkg/dockerfile/dockerfile_test.go
+++ b/pkg/dockerfile/dockerfile_test.go
@@ -238,7 +238,10 @@ func Test_GetOnBuildInstructions(t *testing.T) {
 			if len(cmds) != len(test.expCommands) {
 				t.Fatalf("Expected %d commands, got %d", len(test.expCommands), len(cmds))
 			}
-			resolveCrossStageCommands(cmds, test.stageToIdx)
+			err = resolveCrossStageCommands(cmds, test.stageToIdx, len(test.stageToIdx))
+			if err != nil {
+				t.Fatalf("resolveCrossStageCommands: %v", err)
+			}
 
 			for i, cmd := range cmds {
 				if reflect.TypeOf(cmd) != reflect.TypeOf(test.expCommands[i]) {

